dqn.py

```diff
@@ -0,0 +1,202 @@
+import torch
+import random
+import torch.nn as nn
+import torch.optim as optim
+import torch.nn.functional as F
+import numpy as np
+import re
+import os
+
+
+# DQN with tanh activation function
+def get_activ(a):
+    if a == 't':
+        return nn.Tanh(), 'nn.Tanh()'
+    elif a == 'r':
+        return nn.ReLU(), 'nn.ReLU()'
+    elif a == 'l':
+        return nn.LeakyReLU(), 'nn.LeakyReLU()'
+    elif a == 'e':
+        return nn.ELU(), 'nn.ELU()'
+    elif a == 's':
+        return nn.Sigmoid(), 'nn.Sigmoid()'
+    else:
+        raise ValueError('Unknown activation function: ' + a)
+
+
+def parse_string(hs: str):
+    # split hs into a list of ints and letters
+    h = re.findall(r'\d+|\D+', hs)
+    # check all odd id are letter
+    assert all([h[i].isalpha() for i in range(0, len(h), 2)])
+    # check all even id are int
+    assert all([h[i].isdigit() for i in range(1, len(h), 2)])
+    activ = [h[i] for i in range(0, len(h), 2)]
+    sizes = [int(h[i]) for i in range(1, len(h), 2)]
+    assert len(activ) == len(sizes)
+    return activ, sizes
+
+
+class DQN(nn.Module):
+
+    def __init__(self, input_dim, hs: str, output_dim):
+        super(DQN, self).__init__()
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        # parse hs
+        activ, sizes = parse_string(hs)
+        print('Build a DQN with hidden layers: ' + hs)
+        self.layers = nn.ModuleList([nn.Linear(input_dim, sizes[0]).to(self.device)])
+        print('Linear(' + str(input_dim) + ', ' + str(sizes[0]) + ')')
+
+        for i in range(0, len(sizes)):
+            activ_layer, name = get_activ(activ[i])
+            self.layers.append(activ_layer.to(self.device))
+            print(name)
+            s_in = sizes[i]
+            s_out = sizes[i + 1] if i + 1 < len(sizes) else output_dim
+            self.layers.append(nn.Linear(s_in, s_out).to(self.device))
+            print('Linear(' + str(s_in) + ', ' + str(s_out) + ')')
+
+    def forward(self, x):
+        for layer in self.layers:
+            x = layer(x)
+        return x
+
+
+# Replay Buffer
+# Buffer using numpy array has better performance than using list
+class ReplayBuffer:
```

Comment:
Using a deque may be easier and take up less space than using arrays. Heres a link to the documentation https://docs.python.org/3/library/collections.html#collections.deque

Reply:
Actually I don't think so. Even though deque has O(1) of append, pop and indexing from **both ends**, but the during the training process, the buffer needs to be randomly sampled, i.e. a random access from the middle, this operation is O(n) for deque, but still O(1) for np.array. So as your buffer size going bigger, the deque's performance reduces a lot. About the memory size, I don't know for sure that np.array takes more memory than deque, there will be some overhead for sure. But in this scale of problem, I would prefer the lower time complexity

Reply:
Great explanation. You are probably right. Very thoughtful runtime analysis of this.